python/cudf/cudf/core/column/column.py

```diff
@@ -489,80 +491,130 @@ def __setitem__(self, key: Any, value: Any):
         """
         Set the value of self[key] to value.
 
-        If value and self are of different types,
-        value is coerced to self.dtype
+        If value and self are of different types, value is coerced to
+        self.dtype. Assumes ``self`` and ``value`` are index-aligned.
         """
 
+        # Normalize value to scalar/column
+        value_normalized: Union[cudf.core.scalar.Scalar, ColumnBase]
+        if is_scalar(value):
+            value_normalized = cudf.Scalar(value, dtype=self.dtype)
+        else:
+            value_normalized = as_column(value).astype(self.dtype)
```

Comment:
I suspect that if you use a ternary `value_normalized = cudf.Scalar(...) if is_scalar(value) else ...` mypy will automatically figure out the Union for you so that you don't have to declare it.